### ISSUE TITLE
Conditional logit estimator theory

### ISSUE THREAD

**Issue by kenneth-rios** [kenrios1993/mixed-panel-logit-default-risk#1]
https://www.stata.com/manuals13/rclogit.pdf#rclogit

NOTE: This theory is based off Stata's conditional logit estimator, **NOT** pylogit's. So this is largely irrelevant for our work, but useful to understand nonetheless. It turns out that we could not feasibly predict default probabilities out-of-sample using Stata's estimator. According to their [clogit postestimation manual](https://www.stata.com/manuals13/rclogitpostestimation.pdf):

> Predicted probabilities for conditional logistic regression must be interpreted carefully. Probabilities are estimated for each group as a whole, **not** for individual observations ... As noted previously, the predicted probabilities are really predicted probabilities for the group as a whole ...Thus, if we want to obtain the predicted probabilities for the estimation sample, it is important that, when we make the calculation, **predictions be restricted to the same sample on which we estimated the data.** 

Let $y_c = (y_{c1}, y_{c2}, \dots, y_{cT_{c-1}}, y_{cT_c})$ be a sequences of 1s and 0s for each country $c = 1, 2, \dots, n$. The number of temporal observations for each country is thus allowed to vary by country. Assuming a logistic function over a linear combination of the predictors, the probability of a obtaining a given value for $y_c$ conditional on $\sum_{t=1}^{T_c} y_{ct} = k_{1c}$ is

$Prob(\mathbf{y_c} | k_{1c}) = \frac{exp(\sum_{t=1}^{T_c} y_{ct}(\gamma_t + x_{ct} \mathbf{\beta}))}{\sum_{\mathbf{d_c} \in S_c} exp(\sum_{t=1}^{T_c} d_{ct}(\gamma_t + x_{ct} \mathbf{\beta}))},$

where $\gamma_t$ models the year-fixed effects for each year $t$. The conditioning on country $c$ precisely controls for time-invariant influences on $y_{ct}$ and can thus be interpreted as modeling country-fixed effects.

The conditional log-likelihood is thus $ln(L(\beta, \gamma_t; x_{ct}, y_{ct})) = \dots $. The benefit of utilizing a conditional logit to account for country-fixed effects as opposed to utilizing a simple logit with country-specific dummies is that coefficients for these dummies estimated via maximum likelihood using a simple logit are prone to bias, particularly with relatively small $T_i$'s.

We choose $\beta, \gamma_t$ by maximizing the log-likelihood function subject to L2 regularization. That is, our maximization problem is $\underset{\beta, \gamma_t}{argmax} \ ln(L(\beta, \gamma_t; x_{ct}, y_{ct})) - \frac{\lambda}{2} (\mathbf{\beta}'\mathbf{\beta} + \sum_{t=1}^{max(T_c)} \gamma_t^2)$.

For ease of notation we now append all the $\gamma_t$'s from $t = 1$ to $t = max(T_c)$ into $\beta$ and relabel the new vector of coefficients as $\bar{\beta}$. To convert the problem into a minimization problem, we also multiply through by -1. The objective function is thus 

$\mathcal{L(\bar{\beta}, \gamma_t)} = -(L(\bar{\beta}, \gamma_t; x_{ct}, y_{ct})) + \frac{\lambda}{2} \mathbf{\bar{\beta}}'\mathbf{\bar{\beta}}$. 

Taking first and second partial derivatives of the objective function with respect to  $\bar{\beta}$ leads to 

i). $\nabla \mathcal{L} = \dots - \lambda \bar{\beta}$

ii). $\mathcal{H}(\mathcal{L}) = \dots - \lambda \mathbf{I}_{p \times p}$

where $p$ is the number of covariates to be estimated in $\bar{\beta}$.

Using the gradient and the Hessian of $\mathcal{L}$ given above, we numerically estimate the true values in $\bar{\beta}$ using the Broyden–Fletcher–Goldfarb–Shanno algorithm (_scipy.optimize.minimize_) using the origin as the initial values until convergence is reached up to a particular tolerance.

**Comment by kenneth-rios** [comment]
See page 2 of pylogit's attached theory manual. ~~This is the interpretation we should use in the theory section of our paper.~~ **With country-specific intercepts that pylogit allows in the _conditional logit_, we will actually be able to predict probabilities out-of-sample.** Then we can also throw in year-fixed effects to control for that, which the pylogit creator told me was OK to do. This is my conversation with him: https://stackoverflow.com/questions/46244095/conditional-logit-for-panel-data-in-python/48470963?noredirect=1#comment93810055_48470963

[pylogit_computation.pdf](https://github.com/kenrios1993/conditional_logit_foreign_risk/files/2614136/pylogit_computation.pdf)

**Comment by kenneth-rios** [comment]
https://www.statalist.org/forums/forum/general-stata-discussion/general/1408793-clogit-with-survey-weights-from-nces-data

Unfortunately, the **conditional** logit estimator turns out to be not so helpful because it conditions to predicting only a single default across countries for each year, which is not realistic at all. We aren't getting an unconditional probability from the conditional logit, but rather something else that's very weird and not helpful. I tested this out on fabricated data in pylogit to confirm.

According to the link above, we pretty much only have two options:

i). use a _mixed logit model_ in pylogit (_melogit_ in Stata) that incorporates some random effects that will allow us to back out unconditional default probabilities. The mixed-effects (random coefficients/error components) model estimates country-specific influences not captured by the data by allowing country-specific coefficients for each covariate to be drawn from a normal distribution. The parameters of the normal distribution are then estimated using MLE. This then allows us to calculate the unconditional probability of default by integrating logits over the density of $\beta$ (using pylogit's _panel_predict()_ function) while controlling for year-fixed effects with year dummies.

So basically, we're going to use the mixed-logit estimator because it actually allows us to calculate _unconditional_ probabilities of default, whereas conditional logits do not.

ii). just use the standard unconditional logit estimator and manually include country dummies to model fixed effects. Yes, there will be some bias, but according to [this paper](http://www.kse.org.ua/uploads/file/Coupe(1).pdf), it's barely significant especially if we have over 8 or 12 years of data per panel. I think we have that much data. We can argue that we have decently long panels that bias is not that serious an issue. We can also include L1 (lasso) regularization to drive some covariates to zero and combat test MSE that way. 

We will use ii) in case the mixed-effects logit model in i) fails to converge for very large numbers of parameters.

**Comment by kenneth-rios** [comment]
Thread locked as theory is defunct. See https://github.com/kenrios1993/mixed_panel_logit_default_risk/issues/3
